netbsd/hid.c

```diff
@@ -0,0 +1,1121 @@
+		char arr[256][16];
+		size_t len;
+		const char *child_dev;
+		char devpath[32];
+		int uhid;
+		struct usb_ctl_report_desc ucrd;
+		int use_ucrd;
+		struct hid_device_info *node;
+
+		parent_dev = udi->udi_devnames[i];
+		if (!is_uhid_parent_device(parent_dev))
+			continue;
+
+		len = 0;
+		walk_device_tree(hed->drvctl, parent_dev, 0, arr, &len, is_uhid_device);
+
+		if (len == 0)
+			continue;
+
+		child_dev = arr[0];
+		strlcpy(devpath, "/dev/", sizeof(devpath));
+		strlcat(devpath, child_dev, sizeof(devpath));
+
+		uhid = open(devpath, O_RDONLY | O_CLOEXEC);
+		if (uhid >= 0) {
+			use_ucrd = (ioctl(uhid, USB_GET_REPORT_DESC, &ucrd) != -1);
+			close(uhid);
+		} else {
+			use_ucrd = 0;
+		}
+
+		node = create_device_info(udi, parent_dev, (use_ucrd) ? &ucrd : NULL);
+		if (!node)
+			continue;
+
+		if (!hed->root) {
+			hed->root = node;
+			hed->end = node;
+		} else {
+			hed->end->next = node;
+			hed->end = node;
+		}
+
+		while (hed->end->next)
+			hed->end = hed->end->next;
+	}
+}
+
+static int set_report(hid_device *dev, const uint8_t *data, size_t length, int report)
+{
+	int res;
+	int device_handle;
+	struct usb_ctl_report ucr;
+
+	if (length < 1) {
+		register_device_error(dev, "report must be greater than 1 byte");
+		return -1;
+	}
+
+	device_handle = dev->report_handles[*data];
+	if (device_handle < 0) {
+		register_device_error_format(dev, "unsupported report id: %hhu", *data);
+		return -1;
+	}
+
+	length--;
+	data++;
+
+	if (length > sizeof(ucr.ucr_data)) {
+		register_device_error_format(dev, "report must be less than or equal to %zu bytes", sizeof(ucr.ucr_data));
+		return -1;
+	}
+
+	ucr.ucr_report = report;
+	memcpy(ucr.ucr_data, data, length);
+
+	res = ioctl(device_handle, USB_SET_REPORT, &ucr);
+	if (res == -1) {
+		register_device_error_format(dev, "ioctl (USB_SET_REPORT): %s", strerror(errno));
+		return -1;
+	}
+
+	return (int) (length + 1);
+}
+
+static int get_report(hid_device *dev, uint8_t *data, size_t length, int report)
+{
+	int res;
+	int device_handle;
+	struct usb_ctl_report ucr;
+
+	if (length < 1) {
+		register_device_error(dev, "report must be greater than 1 byte");
+		return -1;
+	}
+
+	device_handle = dev->report_handles[*data];
+	if (device_handle < 0) {
+		register_device_error_format(dev, "unsupported report id: %hhu", *data);
+		return -1;
+	}
+
+	length--;
+	data++;
+
+	if (length > sizeof(ucr.ucr_data)) {
+		length = sizeof(ucr.ucr_data);
+	}
+
+	ucr.ucr_report = report;
+
+	res = ioctl(device_handle, USB_GET_REPORT, &ucr);
+	if (res == -1) {
+		register_device_error_format(dev, "ioctl (USB_GET_REPORT): %s", strerror(errno));
+		return -1;
+	}
+
+	memcpy(data, ucr.ucr_data, length);
+
+	return (int) (length + 1);
+}
+
+int HID_API_EXPORT HID_API_CALL hid_init(void)
+{
+	const char *locale;
+
+	/* indicate no error */
+	register_global_error(NULL);
+
+	/* Set the locale if it's not set. */
+	locale = setlocale(LC_CTYPE, NULL);
+	if (!locale)
+		setlocale(LC_CTYPE, "");
```

Comment:
can you confirm this is required for this backend and not just a copy-paste?

Reply:
Apparently it isn't. I kept it originally because I was assuming it was required for the multi-byte functions to work correctly on utf8 input like it is with `glibc`.